core/src/main/scala/org/apache/spark/deploy/master/Master.scala

```diff
@@ -278,9 +278,24 @@ private[deploy] class Master(
       execOption match {
         case Some(exec) => {
           val appInfo = idToApp(appId)
+          val oldState = exec.state
           exec.state = state
-          if (state == ExecutorState.RUNNING) { appInfo.resetRetryCount() }
+
+          if (state == ExecutorState.RUNNING) {
+            assert(oldState == ExecutorState.LOADING,
```

Comment:
Does `assert` do much the same as `require` in Scala? just making sure this isn't a Java-style `assert` underneath.

Does this cause the master to fail then if the state transitions happen out of order? I pretty much agree that ignoring it isn't good either, but is this fixing the problem? For example if you get `RUNNING` after `LAUNCHING`, you can assume `LOADING` happened, and ignore subsequent `LOADING` messages instead? or does that make more problems?